services/base.js

```diff
@@ -222,14 +223,34 @@ class BaseService {
 };
 
 class BaseJsonService extends BaseService {
+  static get responseSchema() {
+    throw Error('Subclasses should define a response schema')
+  }
+
+  static validateResponse(json) {
+    const { error, value } = Joi.validate(json, this.responseSchema, {
+      allowUnknown: true,
+      stripUnknown: true,
+    });
+    if (error) {
+      throw new InvalidResponse({
+        prettyMessage: 'invalid json response',
+        underlyingError: error,
+      });
+    } else {
+      return value;
+    }
+  }
+
   async _requestJson(url, options = {}, notFoundMessage) {
     return this._sendAndCacheRequest(url,
       {...{ 'headers': { 'Accept': 'application/json' } }, ...options}
     ).then(
       checkErrorResponse.asPromise(
         notFoundMessage ? { notFoundMessage: notFoundMessage } : undefined
       )
-    ).then(asJson);
+    ).then(asJson
+    ).then(json => this.constructor.validateResponse(json));
```

Comment:
Is there any tradeoff in calling `this.constructor.staticMethod()` vs `ClassName.staticMethod()`, or is it just a stylistic consideration?

Reply:
Yea, it allows subclasses to redefine the class method being invoked, and have the expected thing happen.